### ISSUE TITLE
Lightning+DDP: use_distributed_sampler=True always shuffles data in DDP despite using custom sampler

### ISSUE THREAD

**Issue by dmholtz** [Lightning-AI/pytorch-lightning#21131]
### Bug description

When using PyTorch Lightning Trainer in DDP with `use_distributed_sampler=True` and providing a custom sampler to the dataloader, the sampling order from the custom sampler is neglected. Instead, data is always shuffled.

One can track this issue down to [this line](https://github.com/Lightning-AI/pytorch-lightning/blob/634e6e6d06fced4cacfb9de675fc328ace1a06c4/src/lightning/pytorch/trainer/connectors/data_connector.py#L243) where the custom sampler is wrapped with a `DistributedSamplerWrapper`. However, the kwargs contain `shuffle=True` due to [this line](https://github.com/Lightning-AI/pytorch-lightning/blob/e55650dcebe1cb9c50addab06d292a1704b1d288/src/lightning/pytorch/trainer/connectors/data_connector.py#L490).

When using a custom trainer in combination with fabric as suggested [here](https://github.com/Lightning-AI/pytorch-lightning/tree/master/examples/fabric/build_your_own_trainer), the sampling order from the custom sampler is respected. The reason for this is found [here](https://github.com/Lightning-AI/pytorch-lightning/blob/e55650dcebe1cb9c50addab06d292a1704b1d288/src/lightning/fabric/fabric.py#L398) where fabric does not pass `shuffle=True` as keyword argument.

### What version are you seeing the problem on?

v2.5

### How to reproduce the bug

```python
from torch.utils.data import Dataset, Sampler, DataLoader
import lightning as L
import torch.nn as nn
import torch

class IntegerDataset(Dataset):
    def __init__(self):
        self.data = [i for i in range(100)]

    def __len__(self):
        return len(self.data)

    def __getitem__(self, idx):
        return {"data": torch.tensor([self.data[idx]])}

class InOrderSampler(Sampler):
    def __init__(self, dataset):
        self.dataset = dataset

    def __iter__(self):   
       yield from range(len(self.dataset))

    def __len__(self):
        return len(self.dataset)
    
dataset = IntegerDataset()
sampler = InOrderSampler(dataset)
dataloader = DataLoader(dataset=dataset, batch_size=3, sampler=sampler)

class MyModule(L.LightningModule):

    def __init__(self):
        super().__init__()

        self.layer = nn.Linear(10, 10)

    def training_step(self, batch, batch_idx):
        print(batch)

        input = torch.randn(10, 10, device = self.layer.weight.device)
        output = self.layer(input)
        loss = nn.functional.mse_loss(output, input)
        return loss

    def configure_optimizers(self):
        return torch.optim.Adam(self.parameters(), lr=1e-3)

model = MyModule()

trainer = L.Trainer(
    max_epochs=1,
    use_distributed_sampler=True,
)
trainer.fit(
    model=model,
    train_dataloaders=dataloader,
)
```

### Error messages and logs

The integers from the dataset are not printed in order, but are randomly shuffled.

### Environment

<details>
  <summary>Current environment</summary>

```
pytorch-lightning         2.5.1.post0
torch                     2.7.1
torchmetrics              1.7.0
torchvision               0.22.1
```

</details>

**Comment by bhimrazy** [comment]
I’d like to dig deeper into this issue. Assigning myself.

**Comment by SkafteNicki** [comment]
@bhimrazy just wanted to mention that this seems to be a duplicate of https://github.com/Lightning-AI/pytorch-lightning/issues/20253

**Comment by deependujha** [comment]
Hi @dmholtz, thanks for opening the issue.

From the `Trainer`’s docstring on [use_distributed_sampler](https://github.com/Lightning-AI/pytorch-lightning/blob/b7ca4d365978512ae053d7d8a8c2ddc99894e813/src/lightning/pytorch/trainer/trainer.py#L265-L271):

```md
use_distributed_sampler: Whether to wrap the DataLoader's sampler with
                `torch.utils.data.DistributedSampler`.

                If not specified this is toggled automatically for strategies that require it.
                 
                By default, it will add ``shuffle=True`` for the train sampler and
                ``shuffle=False`` for validation/test/predict samplers.

                If you want to disable this logic, you can pass
                ``False`` and add your own distributed sampler in the dataloader hooks. If ``True`` and a distributed
                sampler was already added, Lightning will not replace the existing one...
```

So in your case you’ll want to:

1. Set `use_distributed_sampler=False` in the `Trainer`.
2. Provide your own sampler implementation that handles the distributed logic.

Here’s a minimal working example you can try out:

```python
from torch.utils.data import Dataset, Sampler, DataLoader
import lightning as L
import torch.nn as nn
import torch

class IntegerDataset(Dataset):
    def __init__(self):
        self.data = [i for i in range(100)]

    def __len__(self):
        return len(self.data)

    def __getitem__(self, idx):
        # print(f"Getting item {idx}")
        return {"data": torch.tensor([self.data[idx]])}

class InOrderDistributedSampler(Sampler):
    def __init__(self, dataset, trainer, shuffle=False):
        self.dataset = dataset
        self.global_rank = trainer.global_rank
        self.world_size = trainer.world_size
        self.shuffle = shuffle

    def __iter__(self):
        indices = list(range(len(self.dataset)))

        if self.shuffle:
            # Optionally shuffle the indices
            g = torch.Generator()
            g.manual_seed(0)
            indices = torch.randperm(len(self.dataset), generator=g).tolist()
        
        # yield from indices: starting from self.global_rank, step by self.world_size
        return iter(indices[self.global_rank::self.world_size])

    def __len__(self):
        return len(self.dataset)



    

class MyModule(L.LightningModule):

    def __init__(self):
        super().__init__()

        self.layer = nn.Linear(10, 10)

    def training_step(self, batch, batch_idx):
        print(f"{self.global_rank=} {batch['data']=}")

        input = torch.randn(10, 10, device = self.layer.weight.device)
        output = self.layer(input)
        loss = nn.functional.mse_loss(output, input)
        return loss

    def configure_optimizers(self):
        return torch.optim.Adam(self.parameters(), lr=1e-3)


if __name__ == "__main__":
    trainer = L.Trainer(
        max_epochs=1,
        use_distributed_sampler=False,
    )

    dataset = IntegerDataset()
    sampler = InOrderDistributedSampler(dataset, trainer)
    dataloader = DataLoader(dataset=dataset, batch_size=3, sampler=sampler)

    model = MyModule()

    trainer.fit(
        model=model,
        train_dataloaders=dataloader,
    )

```

This way, Lightning won’t inject its own `DistributedSampler`, and your sampler fully controls how data is partitioned across ranks.

Thanks.

cc: @bhimrazy

**Comment by dmholtz** [comment]
@deependujha Thank you very much for this detailed response, this helps.

Initially, it seems a bit counter-intuitive to obtain rank and world_size from the trainer when building the sampler, but it works.

Since the shuffling behavior is apparently consistent with the docstring, I think this issue can be closed.

**Comment by mojtababahrami** [comment]
I have this same issue, but I am not really satisfied by writing a DistributedSampler from scratch. I expect Lightning to turn shuffle off when a sampler is provided. I don't really know why shuffle should be set to True if a sampler is given as input.

**Comment by deependujha** [comment]
Hi @mojtababahrami, did you tried with `trainer = L.Trainer(..., use_distributed_sampler=False)`, and passed your sampler?

If it doesn't work as expected, can you please share a minimal reproducible code with the expected behaviour?

**Comment by mojtababahrami** [comment]
Yes, I am now using a distributed sampler and set use_distributed_sampler=False but I wish I didn't have to! So let me explain: 
Before starting MultiGPU, I had a simple sampler that was working ok on single GPU, no DDP. But then, starting with DDP, when I pass my simple sampler to the Lightning, it wraps it with DistributedSamplerWrapper and ignores my sampler's sample ordering by setting the shuffle to true. This took me a while to find out the bug, and then, when I did, I replaced my sampler with a distributed version of it, which then worked ok. But the whole thing was very unnecessary and time-consuming. The expected behvaiour is that Lightning converts a sampler to a distributed one respecting the original sampler' order.

**Comment by deependujha** [comment]
Hi @mojtababahrami, can you create a new issue for this?

**Comment by mojtababahrami** [comment]
@deependujha 
I opened this: #21447

**Comment by deependujha** [comment]
Hi @mojtababahrami @dmholtz,

I’ve opened a PR that allows custom samplers to opt out of the Trainer’s automatic shuffling during training. Could you please take a look when you get a chance? #21449

**Comment by dmholtz** [comment]
> I expect Lightning to turn shuffle off when a sampler is provided. I don't really know why shuffle should be set to True if a sampler is given as input.

This is exactly my point with this issue.

**Comment by mojtababahrami** [comment]
@deependujha 
This PR exactly does the job and works for sure. But just from the user perspective, I am still wondering if having `shuffle=True` makes sense at all to have in the first place. I can not think of a use case where a custom sampler is provided, and it needs the Trainer to ignore its ordering. I think keeping it will cause confusion.
However, even if there might be such a use case, I would strongly recommend not making it the default. It is now the default in the PR:
`disable_auto_shuffle = getattr(sampler, "disable_auto_shuffle", False)`

**Comment by deependujha** [comment]
Hi @mojtababahrami, thanks for the thoughtful feedback. I agree with the intuition that if a custom sampler is provided, its ordering should generally be respected.

The main challenge I ran into is that Lightning doesn’t have a reliable way to tell whether a sampler is “custom” versus implicitly created by the DataLoader. When no sampler is passed, PyTorch automatically inserts either `SequentialSampler` or `RandomSampler` based on the `shuffle` flag. From the Trainer’s perspective, both cases look the same: there is always a sampler present.

We could try heuristics, for example checking whether the sampler is an instance of known built-ins (`SequentialSampler`, `RandomSampler`, `SubsetRandomSampler`, `WeightedRandomSampler`, etc.), but this breaks down quickly since custom samplers can subclass these. At that point, we’re guessing intent based on type, which feels fragile.

I also worry about changing the default behavior to “never shuffle when a sampler is present.” Many users don’t explicitly pass a sampler (and end up with a `SequentialSampler`) but still expect training-time shuffling to happen. Even if this wouldn’t be strictly breaking, it would likely be surprising.

Given those constraints, I saw two viable options:

- an explicit opt-in/opt-out mechanism (what this PR does via a sampler attribute), or
- some global toggle (e.g. environment variable), which felt heavier and less discoverable.

I’m very open to a better approach here if you see one, or if you think the default behavior should be reconsidered despite the ambiguity. Happy to adjust the direction based on consensus.

Thanks again for the feedback.

**Comment by mojtababahrami** [comment]
Thanks a lot for following this up!
The issue is that this is not the default behaviour with the non-DDP setup. I mean if I use Lightning on a single GPU (no DDP) and create a Pytorch dataloader without passing any sampler or shuffle argument to it as you mentioned, it will be a sequential dataloader based on [Pytorch Docs](https://docs.pytorch.org/docs/stable/data.html#torch.utils.data.DataLoader), and Lightning will not add shuffling automatically for me in this case. Is that right?

If it does add shuffling for the single GPU case too (which I do not think so) then I am convinced for sure.

**Comment by deependujha** [comment]
https://github.com/Lightning-AI/pytorch-lightning/blob/027455bcd9433f201b1136f68d54b1a07588abe3/src/lightning/pytorch/trainer/connectors/data_connector.py#L177-L194

yeah, you're correct. We update dataloader only when we need to wrap it with `DistributedSampler`. lol :)

**Comment by mojtababahrami** [comment]
Yes that's it! :)
Between these two different behaviours in ddp and non-ddp, I think the non-ddp's way of respecting user's configuration and keeping it untouched makes much more sense and DDP should also be like this. It is not really Lightning's job to decide shuffling on behalf of the user. 
I think this was even the intention of whoever wrote the DistributedSamplerWrapper based on its Docstring:

https://github.com/Lightning-AI/pytorch-lightning/blob/027455bcd9433f201b1136f68d54b1a07588abe3/src/lightning/fabric/utilities/distributed.py#L359-L364
